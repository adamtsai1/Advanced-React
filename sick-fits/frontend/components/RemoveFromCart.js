import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const BigButton = styled.button`
    font-size: 3rem;
    background: none;
    border: none;
    &:hover {
        color: ${props => props.theme.red};
        cursor: pointer;
    }
`;

import { CURRENT_USER_QUERY } from './User';

const DELETE_CART_ITEM_MUTATION = gql`
    mutation DELETE_CART_ITEM_MUTATION($id: ID!) {
        removeFromCart(id: $id) {
            id
        }
    }
`;

class RemoveFromCart extends Component {
    static propTypes = {
        id: PropTypes.string.isRequired,
    };

    // This gets called as soon as we get a response from the server after a mutation has been performed.
    // You can use this instead of refetchQueries.
    update = (cache, payload) => {

        // 1. Read the cache
        const data = cache.readQuery({ query: CURRENT_USER_QUERY });

        // 2. Remove that item from the cart
        const cartItemId = payload.data.removeFromCart.id;
        data.me.cart = data.me.cart.filter(cartItem => cartItem.id !== cartItemId);

        // 3. Write it back to the cache
        cache.writeQuery({ query: CURRENT_USER_QUERY, data });
    };

    render() {
        return (
            <Mutation
                mutation={DELETE_CART_ITEM_MUTATION}
                variables={{ id: this.props.id }}
                update={this.update}
                optimisticResponse={{
                    __typename: 'Mutation',
                    removeFromCart: {
                        __typename: 'CartItem',
                        id: this.props.id,
                    },
                }}
            >
                {(deleteCartItem, { loading, }) => (
                    <BigButton
                        disabled={loading}
                        title="Delete Item"
                        onClick={() => {
                            deleteCartItem().catch(error => alert(error.message));
                        }}
                    >&times;</BigButton>
                )}
            </Mutation>
        );
    }
}

export default RemoveFromCart;
