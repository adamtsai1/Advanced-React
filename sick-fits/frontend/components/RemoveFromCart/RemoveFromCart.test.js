import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import MockedProvider from 'react-apollo/test-utils';
import RemoveFromCart, { DELETE_CART_ITEM_MUTATION } from './';

describe('<RemoveFromCart />', () => {
    it('renders and matches snapshot', () => {
        expect(1).toBe(1);
    });
});
