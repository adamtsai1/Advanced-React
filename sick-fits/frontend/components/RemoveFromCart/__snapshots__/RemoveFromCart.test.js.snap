// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RemoveFromCart /> renders and matches snapshot 1`] = `
<button
  className="RemoveFromCart__BigButton-lmnsq8-0 hhAILj"
  disabled={false}
  onClick={[Function]}
  title="Delete Item"
>
  Ã—
</button>
`;
